### PR TITLE
Feature/custom fields background

### PR DESCRIPTION
**Pull request by Maxell92**
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Automated tests included? | Y
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | N
| Deprecations? | Y

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:

This PR adds an option to run ALTER TABLE statement on leads table in background. There is a new option in settings for this.

#### Steps to test this PR:
1. Checkout this PR, clear cache and test that adding / removing CS works properly. Check that `published` option works correctly for creating a CS.
2. Add an option `'create_custom_field_in_background' => true,` to your `local.php`
3. Clear cache
4. Add a new custom field. You should get a notification that custom field will be created and custom field should be unpublished. Test there is no way to publish this field.
5. Run command: `app/console mautic:custom-field:create-column -i 114 --user=1 --env=dev`. `114` is ID of new CF, `1` is ID of user which will get a notification. Should be ID of your user.
6. Reload page with CF list. Field should be published / unpublished (depends on your choice when creating CF, try both variants). There should be a notification in notification center that CF was created.
7) Export (backup) your `leads` table and use this one:
```
CREATE TABLE `leads` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `owner_id` int(11) DEFAULT NULL,
  `stage_id` int(11) DEFAULT NULL,
  `is_published` tinyint(1) NOT NULL,
  `date_added` datetime DEFAULT NULL COMMENT '(DC2Type:datetime)',
  `created_by` int(11) DEFAULT NULL,
  `created_by_user` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `date_modified` datetime DEFAULT NULL COMMENT '(DC2Type:datetime)',
  `modified_by` int(11) DEFAULT NULL,
  `modified_by_user` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `checked_out` datetime DEFAULT NULL COMMENT '(DC2Type:datetime)',
  `checked_out_by` int(11) DEFAULT NULL,
  `checked_out_by_user` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `points` int(11) NOT NULL,
  `last_active` datetime DEFAULT NULL COMMENT '(DC2Type:datetime)',
  `internal` longtext COLLATE utf8_unicode_ci DEFAULT NULL COMMENT '(DC2Type:array)',
  `social_cache` longtext COLLATE utf8_unicode_ci DEFAULT NULL COMMENT '(DC2Type:array)',
  `date_identified` datetime DEFAULT NULL COMMENT '(DC2Type:datetime)',
  `preferred_profile_image` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `title` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `firstname` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `lastname` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `company` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `position` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `email` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `phone` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `mobile` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `address1` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `address2` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `city` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `state` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `zipcode` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `timezone` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `country` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `fax` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `preferred_locale` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `attribution_date` datetime DEFAULT NULL COMMENT '(DC2Type:datetime)',
  `attribution` double DEFAULT NULL,
  `website` longtext COLLATE utf8_unicode_ci DEFAULT NULL,
  `facebook` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `foursquare` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `googleplus` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `instagram` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `linkedin` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `skype` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `twitter` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_1` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_2` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_3` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_4` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_5` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_7` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_6` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_8` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_10` longtext COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_11` longtext COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_12` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_13` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_14` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_15` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_16` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_17` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_18` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_19` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_20` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_21` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_22` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_23` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_24` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_25` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_26` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_27` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_28` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_29` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_30` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_31` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_32` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_33` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_34` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_35` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_36` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_37` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_38` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_39` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_40` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_41` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_42` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_43` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_44` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_45` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_46` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_47` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_48` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_49` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_50` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_51` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_52` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_53` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_54` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_55` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_56` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_57` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_58` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_59` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  `test_field_60` varchar(255) COLLATE utf8_unicode_ci DEFAULT NULL,
  PRIMARY KEY (`id`),
  KEY `IDX_179045527E3C61F9` (`owner_id`),
  KEY `IDX_179045522298D193` (`stage_id`),
  KEY `lead_date_added` (`date_added`),
  KEY `title_search` (`title`),
  KEY `firstname_search` (`firstname`),
  KEY `lastname_search` (`lastname`),
  KEY `company_search` (`company`),
  KEY `position_search` (`position`),
  KEY `email_search` (`email`),
  KEY `mobile_search` (`mobile`),
  KEY `phone_search` (`phone`),
  KEY `points_search` (`points`),
  KEY `fax_search` (`fax`),
  KEY `address1_search` (`address1`),
  KEY `address2_search` (`address2`),
  KEY `city_search` (`city`),
  KEY `state_search` (`state`),
  KEY `zipcode_search` (`zipcode`),
  KEY `country_search` (`country`),
  KEY `preferred_locale_search` (`preferred_locale`),
  KEY `attribution_date_search` (`attribution_date`),
  KEY `attribution_search` (`attribution`),
  KEY `facebook_search` (`facebook`),
  KEY `foursquare_search` (`foursquare`),
  KEY `googleplus_search` (`googleplus`),
  KEY `instagram_search` (`instagram`),
  KEY `linkedin_search` (`linkedin`),
  KEY `skype_search` (`skype`),
  KEY `twitter_search` (`twitter`),
  KEY `contact_attribution` (`attribution`,`attribution_date`),
  KEY `date_added_country_index` (`date_added`,`country`),
  KEY `date_identified` (`date_identified`),
  KEY `test_field_1_search` (`test_field_1`),
  KEY `test_field_2_search` (`test_field_2`),
  KEY `test_field_3_search` (`test_field_3`),
  KEY `test_field_4_search` (`test_field_4`),
  KEY `test_field_5_search` (`test_field_5`),
  KEY `test_field_7_search` (`test_field_7`),
  KEY `test_field_6_search` (`test_field_6`),
  KEY `test_field_8_search` (`test_field_8`),
  KEY `test_field_12_search` (`test_field_12`),
  KEY `test_field_13_search` (`test_field_13`),
  KEY `test_field_14_search` (`test_field_14`),
  KEY `test_field_15_search` (`test_field_15`),
  KEY `test_field_16_search` (`test_field_16`),
  KEY `test_field_17_search` (`test_field_17`),
  KEY `test_field_18_search` (`test_field_18`),
  KEY `test_field_19_search` (`test_field_19`),
  KEY `test_field_20_search` (`test_field_20`),
  KEY `test_field_21_search` (`test_field_21`),
  KEY `test_field_22_search` (`test_field_22`),
  KEY `test_field_23_search` (`test_field_23`),
  KEY `test_field_24_search` (`test_field_24`),
  KEY `test_field_25_search` (`test_field_25`),
  KEY `test_field_26_search` (`test_field_26`),
  KEY `test_field_27_search` (`test_field_27`),
  KEY `test_field_28_search` (`test_field_28`),
  KEY `test_field_29_search` (`test_field_29`),
  KEY `test_field_30_search` (`test_field_30`),
  KEY `test_field_31_search` (`test_field_31`),
  KEY `test_field_32_search` (`test_field_32`),
  KEY `test_field_33_search` (`test_field_33`),
  KEY `test_field_34_search` (`test_field_34`),
  CONSTRAINT `FK_179045522298D193` FOREIGN KEY (`stage_id`) REFERENCES `stages` (`id`) ON DELETE SET NULL,
  CONSTRAINT `FK_179045527E3C61F9` FOREIGN KEY (`owner_id`) REFERENCES `users` (`id`) ON DELETE SET NULL
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci;
```
Try to create a CF. You should get a notification that CF cannot be created.

#### List deprecations along with the new alternative:
1. FieldModel::getFieldList() > FieldList::getFieldList()
2. FieldModel::getUniqueIdentifierFields() > FieldsWithUniqueIdentifier::getFieldsWithUniqueIdentifier
3. FieldModel::getSchemaDefinition() > SchemaDefinition::getSchemaDefinition()

### CODE REVIEW

**Comment by Woeler** [comment]
Quick question. Does this solve the issue that MySQL does not allow more than 64 keys on one table?

**Comment by Woeler** [comment]
Would it make sense to drop indexes from custom fields? Let's say you are at the limit, and we add a new core feature that adds a new field to the leads table with an index, this migration will then fail.

**Comment by Maxell92** [comment]
I do not think it is a good idea. Indexes are really important for performance.

**Comment by Woeler** [comment]
Maybe make them optional, as in a yes/no field when creating a CF? that way, this migration error can be mitigated.

**Comment by Maxell92** [comment]
I do not think that customer will know what to choose. So we need to implement some defaults and hiding this. Sounds like a new feature request :)

**Comment by Woeler** [comment]
I for instance would trade many core field indexes for custom fields one if I could. E.g. I have no use for having an index on the Twitter handle field, because I do not use it.

But this should indeed go into another feature. I'm not a fan of hard limits, but this is a good start definitely.

**Comment by Maxell92** [comment]
The purpose of this PR is to add a possibility to make CF change in the background to prevent timing out + refactoring of related code.
We can definitely talk about change for indexes, but I think this should be done in another FR.

**Comment by npracht** [comment]
Ok so if i understand it is a enable/disable feature right ?
And @Woeler would suggest more advanced settings if enabled?

If that is a correct understanding, then i agree too :)

**Comment by Maxell92** [comment]
@npracht Yes, the new behaviour is disabled by default. You need to add config option to allow processing CF add in the background and setup cron (or run command manually).

The index thing is not related. It could be done for both config options.

**Comment by mleffler** [comment]
This has been code reviewed, tested, and tested in production.

**Comment by anton-vlasenko** [comment]
I've resolved conflicts. It can be merged now.

**Comment by alanhartless** [comment]
We have a PR built on top of this that enables configuration options for custom field length and whether the field should be indexable or not. But it requires this PR to be merged first.

**Comment by kuzmany** [comment]
I am going to test it.

File: app/migrations/Version20180921144421.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+{
+    /**
+     * @param Schema $schema
+     *
+     * @throws SkipMigrationException
+     * @throws \Doctrine\DBAL\Schema\SchemaException
+     */
+    public function preUp(Schema $schema)
+    {
+        if ($schema->getTable($this->prefix.'lead_fields')->hasColumn('column_is_not_created')) {
```
Comment (kuzmany) [inline app/migrations/Version20180921144421.php]: Should be hasTable

Failed during migrations

`Migration 20170516000000 failed during Pre-Checks. Error An exception occurred while executing 'SELECT t0.is_published AS is_published_1, t0.date_added AS date_added_2, t0.created_by AS created_by_3, t0.created_by_user AS created_by_user_4, t0.date_modified AS date_modified_5, t0.modified_by AS modified_by_6, t0.modified_by_user AS modified_by_user_7, t0.checked_out AS checked_out_8, t0.checked_out_by AS checked_out_by_9, t0.checked_out_by_user AS checked_out_by_user_10, t0.id AS id_11, t0.label AS label_12, t0.alias AS alias_13, t0.type AS type_14, t0.field_group AS field_group_15, t0.default_value AS default_value_16, t0.is_required AS is_required_17, t0.is_fixed AS is_fixed_18, t0.is_visible AS is_visible_19, t0.is_short_visible AS is_short_visible_20, t0.is_listable AS is_listable_21, t0.is_publicly_updatable AS is_publicly_updatable_22, t0.is_unique_identifer AS is_unique_identifer_23, t0.field_order AS field_order_24, t0.object AS object_25, t0.properties AS properties_26, t0.column_is_not_created AS column_is_not_created_27, t0.original_is_published_value AS original_is_published_value_28 FROM lead_fields t0 WHERE t0.alias = ? LIMIT 1' with params ["points"]:`

**Review by kuzmany — DISMISSED** [review]
I am trying to test it.
But column_is_not_created is always set to 0  and created in leads table even create_custom_field_in_background is set to true. 

I did db schema by command, but expect custom field created with flag column_is_not_created = 1

**Comment by npracht** [comment]
@Maxell92 please see Zdeno's comment and Travis issue. Thanks!

**Comment by kuzmany** [comment]
@Maxell92 I didn't find any use of this method \Mautic\LeadBundle\Field\LeadFieldSaver::saveLeadFieldEntityWithoutColumnCreated

**Comment by npracht** [comment]
Hi there!
It has been decided to not create any extra Mautic 2 minor versions (which means no more features in M2) knowing that Mautic 3 is coming very soon.

**We now want to integrate your contribution in the Mautic 3 roadmap as 3.1.0 candidate.**

## How to do?
1. Check if your feature or enhancement is still missing and relevant in Mautic 3
2. **If Yes**: rebase your PR against the 3.x branch
3. **If No**: please consider closing your PR if you are absolutely sure that the feature has made it into Mautic 3 already

**Please report results by commenting on your PR to make us administration easier.**
______
You can more information on how to do all of that on [this blog post "Getting you PR ready for Mautic 3"](https://www.mautic.org/blog/developer/getting-your-pr-ready-mautic-3).

**Comment by hluchas** [comment]
Rebasing this agains 3.x ...

EDIT:
New PR https://github.com/mautic/mautic/pull/8828

**Pull request closed without merging**